presto-plugin-toolkit/src/main/java/com/facebook/presto/security/RefreshingAccessControlUtils.java

```diff
@@ -0,0 +1,102 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.security;
+
+import com.facebook.presto.spi.connector.ConnectorAccessControl;
+import com.facebook.presto.spi.connector.DelegateConnectorAccessControl;
+import com.facebook.presto.spi.security.DelegateSystemAccessControl;
+import com.facebook.presto.spi.security.SystemAccessControl;
+import io.airlift.units.Duration;
+
+import java.nio.file.Path;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+public final class RefreshingAccessControlUtils
+{
+    private RefreshingAccessControlUtils()
+    {
+    }
+
+    public static SystemAccessControl refreshingSystemAccessControl(Supplier<SystemAccessControl> factory, Duration refreshPeriod)
+    {
+        return new DelegateSystemAccessControl(new Refresher<>(factory, () -> true, refreshPeriod));
+    }
+
+    public static SystemAccessControl refreshingSystemAccessControl(Supplier<SystemAccessControl> factory, Supplier<Boolean> refreshCondition, Duration refreshPeriod)
+    {
+        return new DelegateSystemAccessControl(new Refresher<>(factory, refreshCondition, refreshPeriod));
+    }
+
+    public static ConnectorAccessControl refreshingConnectorAccessControl(Supplier<ConnectorAccessControl> factory, Duration refreshPeriod)
+    {
+        return new DelegateConnectorAccessControl(new Refresher<>(factory, () -> true, refreshPeriod));
+    }
+
+    public static ConnectorAccessControl refreshingConnectorAccessControl(Supplier<ConnectorAccessControl> factory, Supplier<Boolean> refreshCondition, Duration refreshPeriod)
+    {
+        return new DelegateConnectorAccessControl(new Refresher<>(factory, refreshCondition, refreshPeriod));
+    }
+
+    public static Supplier<Boolean> wasFileModified(Path path)
+    {
+        AtomicLong configFileNameLastModification = new AtomicLong();
+        return () -> {
+            long lastModification = path.toFile().lastModified();
+            return configFileNameLastModification.get() != lastModification &&
+                    configFileNameLastModification.compareAndSet(configFileNameLastModification.get(), lastModification);
+        };
+    }
+
+    private static final class Refresher<T>
+            implements Supplier<T>
+    {
+        private final Supplier<T> supplier;
+        private final Supplier<Boolean> refreshCondition;
```

Comment:
it strucked me -- why not simplify this all down to `com.google.common.base.Suppliers#memoizeWithExpiration`?

i know we lose "refresh on file change" but we gain simplicity. Maybe this is worth it?